xmodule/tests/test_item_bank.py

```diff
@@ -0,0 +1,521 @@
+        child = self.item_bank.get_child_blocks()[0]
+        event_data = self._assert_event_was_published("assigned")
+        block_info = {
+            "usage_key": str(child.usage_key),
+        }
+        assert event_data ==\
+               {'location': str(self.item_bank.usage_key),
+                'added': [block_info],
+                'result': [block_info],
+                'previous_count': 0, 'max_count': 1}
+        self.publisher.reset_mock()
+
+        # Now increase max_count so that one more child will be added:
+        self.item_bank.max_count = 2
+        children = self.item_bank.get_child_blocks()
+        assert len(children) == 2
+        child, new_child = children if children[0].usage_key == child.usage_key else reversed(children)
+        event_data = self._assert_event_was_published("assigned")
+        assert event_data['added'][0]['usage_key'] == str(new_child.usage_key)
+        assert len(event_data['result']) == 2
+        assert event_data['previous_count'] == 1
+        assert event_data['max_count'] == 2
+
+    def test_removed_overlimit(self):
+        """
+        Test the "removed" event emitted when we un-assign blocks previously assigned to a student.
+        We go from one blocks assigned to none because max_count has been decreased.
+        """
+        # Decrease max_count to 1, causing the block to be overlimit:
+        self.item_bank.get_child_blocks()  # This line is needed in the test environment or the change has no effect
+        self.publisher.reset_mock()  # Clear the "assigned" event that was just published.
+        self.item_bank.max_count = 0
+
+        # Check that the event says that one block was removed, leaving no blocks left:
+        children = self.item_bank.get_child_blocks()
+        assert len(children) == 0
+        event_data = self._assert_event_was_published("removed")
+        assert len(event_data['removed']) == 1
+        assert event_data['result'] == []
+        assert event_data['reason'] == 'overlimit'
+
+    @ddt.data(
+        (
+            [("problem", "My_Item_1"), ("problem", "My_Item_2")],
+            ("problem", "My_Item_1"),
+            [("problem", "My_Item_2"), ("problem", "My_Item_3")],
+        ),
+        (
+            [("problem", "My_Item_1"), ("problem", "My_Item_2")],
+            ("problem", "My_Item_2"),
+            [("problem", "My_Item_1"), ("problem", "My_Item_3")],
+        ),
+        (
+            [("problem", "My_Item_3"), ("problem", "My_Item_0")],
+            ("problem", "My_Item_3"),
+            [("problem", "My_Item_2"), ("problem", "My_Item_3")],
+        ),
+        (
+            [("problem", "My_Item_3"), ("problem", "My_Item_0")],
+            ("problem", "My_Item_0"),
+            [("problem", "My_Item_3"), ("problem", "My_Item_2")],
+        ),
+    )
+    @ddt.unpack
+    def test_removed_invalid(self, to_select_initial, to_drop, to_select_new):
```

Comment:
FYI @ormsbee , @bradenmacdonald -- I'm not proud of how long and convoluted this test case is, but I feel very confident that it thoroughly tests the selection & re-selection process, which has consistently shown itself to be the most complex aspect of this block. I'm willing to break it into smaller + more readable test cases in a follow-up, but I don't think it's important right now.